datafusion/optimizer/src/optimizer.rs

```diff
@@ -451,6 +468,33 @@ impl Optimizer {
     }
 }
 
+/// These are invariants to hold true for each logical plan.
+/// Do necessary check and fail the invalid plan.
+///
+/// Checks for elements which are immutable across optimizer passes.
+fn check_plan(
+    check_name: &str,
+    plan: &LogicalPlan,
+    prev_schema: Arc<DFSchema>,
+) -> Result<()> {
+    // verify invariant: optimizer rule didn't change the schema
+    assert_schema_is_the_same(check_name, &prev_schema, plan)?;
+
+    // verify invariant: fields must have unique names
+    assert_unique_field_names(plan)?;
+
+    /* This current fails for:
+       - execution::context::tests::cross_catalog_access
+       - at test_files/string/string.slt:46
+               External error: query failed: DataFusion error: Optimizer rule 'eliminate_nested_union' failed
+    */
+    // verify invariant: equivalent schema across union inputs
+    // assert_unions_are_valid(check_name, plan)?;
+
+    // TODO: trait API and provide extension on the Optimizer to define own validations?
```

Comment:
Here is a mention of the extensibility of invariants. Options include:
* for general invariants:
   * defined as being checked before/after each OptimizerRule, and applied here in `check_plan()` (or equivalent code)
   * we could provide `Optimizer.invariants = Vec<Arc<dyn InvariantCheck>>` for user-defined invariants
* for invariants specific for a given OptimizerRule:
   * we could provide `OptimizerRule::check_invariants()` such that certain invariants are only checked for a given rule (instead of all rules)
   * for a user-defined OptimizerRule, users can also check their own invariants

Ditto for the AnalyzerRule passes. Altho I wasn't sure about how much is added complexity and planning time overhead - as @Omega359 mentions we could [make it configurable](https://github.com/apache/datafusion/pull/13651#issuecomment-2519006923) (e.g. run for CI and debugging in downstream projects). 

This WIP is about proposing different ideas of what we could do. 🤔

Reply:
Maybe it can be controlled through environment variables, similar to `RUST_LOG`. Enable it for debugging when problems are encountered or during an upgrade.

Reply:
do you have example use-case for user-defined plan invariants?

Reply:
We have some special invariants for our `SortPreservingMerge` replacement, `ProgressiveEval` (related to time ranges of parquet files) that would be great to be able to encode

Reply:
> Maybe it can be controlled through environment variables, similar to RUST_LOG or RUST_BACKTRACE. Enable it for debugging when problems are encountered or during an upgrade.

We could also add it as a `debug_assert!` after each optimizer pass and call the real validation
1. After analyze
2. After all the optimizer passes

Reply:
> We have some special invariants for our `SortPreservingMerge` replacement, `ProgressiveEval` (related to time ranges of parquet files) that would be great to be able to encode

is this about `LogicalPlan::Extension`? I agree it makes sense to support validation of these if we validate the overall plan.

Reply:
I've added this to the followup items list: https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777